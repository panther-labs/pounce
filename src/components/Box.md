Responsive box-model layout component.

```jsx harmony
import Text from 'components/Text';

<Box p={9} fontSize={4} width={300} color="white" bg="primary200">
  <Text size="medium" color="white" fontSize={4}>
    Box
  </Text>
</Box>;
```
