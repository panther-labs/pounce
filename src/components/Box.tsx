import React from 'react';
import styled, {
  CSSObject,
  FlattenSimpleInterpolation,
  FlattenInterpolation,
} from 'styled-components';
import * as StyledSystem from 'styled-system';
import { Theme } from 'styled-system';

export interface BoxProps
  extends StyledSystem.SpaceProps,
    StyledSystem.ColorProps,
    StyledSystem.FontFamilyProps,
    StyledSystem.FontSizeProps,
    StyledSystem.FontWeightProps,
    StyledSystem.WidthProps,
    StyledSystem.MinWidthProps,
    StyledSystem.HeightProps,
    StyledSystem.GridRowProps,
    StyledSystem.GridColumnProps,
    StyledSystem.ShadowProps,
    StyledSystem.BorderProps,
    StyledSystem.PositionProps,
    StyledSystem.FlexProps {
  /** The native HTML element to use for this component.
   * @default "div"
   * */
  as?: React.ElementType;

  /** Additional custom inline CSS to pass to the element
   * @default "{}"
   * */
  css?: CSSObject | FlattenSimpleInterpolation | FlattenInterpolation<Theme> | string;
}

/** Responsive box-model layout component. Apart from the defined props,
 * it also supports all the native HTML attributes. */
const Box: React.FC<BoxProps> = styled.div`
  ${StyledSystem.space}
  ${StyledSystem.color}
  ${StyledSystem.fontFamily}
  ${StyledSystem.fontSize}
  ${StyledSystem.fontWeight}
  ${StyledSystem.width}
  ${StyledSystem.minWidth}
  ${StyledSystem.height}
  ${StyledSystem.gridRow}
  ${StyledSystem.gridColumn}
  ${StyledSystem.shadow}
  ${StyledSystem.border}
  ${StyledSystem.position}
  ${StyledSystem.flex}
`;

export default Box;
