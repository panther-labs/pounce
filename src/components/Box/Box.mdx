Responsive box-model layout component.

The Box component is the primary component that powers the entire library. It accepts as props
almost all CSS properties and automatically converts theme values to actual browser-friendly values.

For example you can say `fontSize="medium"` and it will know to translate it to proper rem/px values.

A Box with the 9th value on the spacing scale, with large font and blue background:

```typescript jsx
<Box p={9} width={300} bg="blue-400" fontSize="large">
  Box
</Box>
```

A Box with the red font color and huge font-size:

```typescript jsx
<Box color="red-300" fontSize="2x-large">
  Box
</Box>
```

A Box that has a responsive width according to the screen size:

```typescript jsx
<Box p={9} bg="red-200" width={[300, 600, 900]} fontSize="2x-large">
  Box
</Box>
```

A Box can also implement other components under the hood and accept its props:

```typescript jsx
<Box as={Button} variantColor="pink-700">
  Box
</Box>
```
