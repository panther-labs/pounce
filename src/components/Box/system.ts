import { createShouldForwardProp, props } from '@styled-system/should-forward-prop';
import { SerializedStyles } from '@emotion/react';
import * as StyledSystem from 'styled-system';
import { Theme } from '../../theme';
import React from 'react';

// prettier-ignore
// Gather all the styled-system props that we are going to pass
type ThemedStyleProps =
  StyledSystem.SpaceProps<Theme, React.CSSProperties['margin']> &
  StyledSystem.ColorProps<Theme> &
  StyledSystem.LayoutProps<Theme> &
  StyledSystem.BackgroundProps<Theme> &
  StyledSystem.GridProps<Theme> &
  StyledSystem.ShadowProps<Theme> &
  StyledSystem.BorderProps<Theme> &
  StyledSystem.PositionProps<Theme> &
  StyledSystem.FlexboxProps<Theme> &
  StyledSystem.TypographyProps<Theme>;

// Props related to Link components in @reach/router & react-router
type RoutingProps = {
  /** The url path to navigate, if the component implements a Link
   * @ignore
   */
  to?: string;
};

// Props related to the usage of the Emotion CSS-in-JS library
type EmotionProps = {
  /** The React Component or native HTML element to render instead.
   * @default "div"
   * @ignore
   */
  is?: React.ElementType;
  /** Additional custom inline CSS to pass to the element
   * @ignore
   */
  css?: SerializedStyles;
};

// Gather the custom-named props that styled-system should accept
export interface CustomStyleProps {
  shadow?: ThemedStyleProps['boxShadow'];
  fill?: ThemedStyleProps['color'];
  stroke?: ThemedStyleProps['color'];
  textDecoration?: StyledSystem.ResponsiveValue<React.CSSProperties['textDecoration']>;
  overflowX?: StyledSystem.OverflowProps['overflow'];
  overflowY?: StyledSystem.OverflowProps['overflow'];
  textTransform?: StyledSystem.ResponsiveValue<React.CSSProperties['textTransform']>;

  animation?: StyledSystem.ResponsiveValue<React.CSSProperties['animation']>;
  transform?: StyledSystem.ResponsiveValue<React.CSSProperties['transform']>;
  visibility?: StyledSystem.ResponsiveValue<React.CSSProperties['visibility']>;
  whiteSpace?: StyledSystem.ResponsiveValue<React.CSSProperties['whiteSpace']>;
  pointerEvents?: StyledSystem.ResponsiveValue<React.CSSProperties['pointerEvents']>;
  wordBreak?: StyledSystem.ResponsiveValue<React.CSSProperties['wordBreak']>;
  overflowWrap?: StyledSystem.ResponsiveValue<React.CSSProperties['overflowWrap']>;
  textOverflow?: StyledSystem.ResponsiveValue<React.CSSProperties['textOverflow']>;
  cursor?: StyledSystem.ResponsiveValue<React.CSSProperties['cursor']>;
  resize?: StyledSystem.ResponsiveValue<React.CSSProperties['resize']>;
  transition?: StyledSystem.ResponsiveValue<React.CSSProperties['transition']>;
  objectFit?: StyledSystem.ResponsiveValue<React.CSSProperties['objectFit']>;
  outline?: StyledSystem.ResponsiveValue<React.CSSProperties['outline']>;
  willChange?: StyledSystem.ResponsiveValue<React.CSSProperties['willChange']>;
}

export const customStyleProps: Record<keyof CustomStyleProps, object | boolean> = {
  shadow: {
    property: 'boxShadow' as const,
    scale: 'shadows',
  },
  fill: {
    property: 'fill' as const,
    scale: 'colors',
  },
  stroke: {
    property: 'stroke' as const,
    scale: 'colors',
  },
  textDecoration: true,
  overflowX: true,
  overflowY: true,
  textTransform: true,
  animation: true,
  transform: true,
  visibility: true,
  whiteSpace: true,
  pointerEvents: true,
  wordBreak: true,
  overflowWrap: true,
  textOverflow: true,
  cursor: true,
  resize: true,
  transition: true,
  objectFit: true,
  outline: true,
  willChange: true,
};

// All of the allowed props gathered together
export type SystemProps = ThemedStyleProps & CustomStyleProps & RoutingProps & EmotionProps;

// extend the forwarded props by stuff that styled-system doesn't deal with
export const shouldForwardProp = createShouldForwardProp([
  ...props,
  'textDecoration',
  'pointerEvents',
  'visibility',
  'transform',
  'cursor',
]);
