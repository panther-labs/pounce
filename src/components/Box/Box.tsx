import React from 'react';
import { pounce, NativeAttributes, PounceComponentProps } from 'system';
import { __DEV__ } from 'utils/helpers';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type BoxProps<T extends React.ElementType = any> = PounceComponentProps<T>;
export type { NativeAttributes };

/**
 * Responsive box-model layout component. Apart from the defined props,
 * it also supports all the native HTML attributes.
 * */
const Box = pounce('div');

// FIXME: This overrides the fact that the box is a `div`. Components implementing this will warn us
//  that they can't implement `<Box>` when they themselves are not a div. We should ideally fix that
//  and have `ref` can be properly typed  in all components
export default Box as React.FC<BoxProps>;

if (__DEV__) {
  Box.displayName = 'Box';
}
