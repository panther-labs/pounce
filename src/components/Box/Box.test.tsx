import React from 'react';
import { render } from 'test-utils';
import Box from './Box';

describe('Box', () => {
  it('renders', async () => {
    const { container } = await render(<Box />);
    expect(container).toMatchSnapshot();
  });

  it('allows truncated values', async () => {
    const { container } = await render(<Box truncated />);
    expect(container).toMatchSnapshot();
  });

  it('allows specifying an opacity for the background color', async () => {
    const { container } = await render(
      <React.Fragment>
        <Box backgroundColor="white" backgroundOpacity="0.5" />
        <Box backgroundColor="white" backgroundOpacity={0.5} />
        <Box backgroundColor="white" backgroundOpacity="50" />
        <Box backgroundColor="white" backgroundOpacity={50} />
      </React.Fragment>
    );
    expect(container).toMatchSnapshot();
  });

  it('allows specifying an opacity for the border color', async () => {
    const { container } = await render(
      <React.Fragment>
        <Box border="1px solid" borderColor="white" borderOpacity="0.5" />
        <Box border="1px solid" borderColor="white" borderOpacity={0.5} />
        <Box border="1px solid" borderColor="white" borderOpacity="50" />
        <Box border="1px solid" borderColor="white" borderOpacity={50} />
      </React.Fragment>
    );
    expect(container).toMatchSnapshot();
  });

  it('allows styling on demand', async () => {
    const { container } = await render(
      <Box color="black" background="red" textTransform="uppercase" />
    );
    expect(container).toMatchSnapshot();
  });
});
