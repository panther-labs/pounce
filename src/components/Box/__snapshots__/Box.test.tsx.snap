// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box allows specifying an opacity for the background color 1`] = `
.pounce-0 {
  background-color: white;
  background-color: rgba(255,255,255,0.5);
}

<div>
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Box allows specifying an opacity for the border color 1`] = `
.pounce-0 {
  border: 1px solid;
  border-color: white;
  border-color: rgba(255,255,255,0.5);
}

<div>
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Box allows styling on demand 1`] = `
.pounce-0 {
  color: black;
  background: red;
  text-transform: uppercase;
}

<div>
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Box allows truncated values 1`] = `
.pounce-0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div>
  <div
    class="pounce-0"
  />
</div>
`;

exports[`Box renders 1`] = `
<div>
  <div
    class="pounce-0"
  />
</div>
`;
